src/manager/node.c

```diff
@@ -1082,6 +1091,13 @@ static int agent_request_callback(sd_bus_message *m, void *userdata, UNUSED sd_b
         return req->cb(req, m, ret_error);
 }
 
+int agent_request_cancel(AgentRequest *r) {
+        _cleanup_agent_request_ AgentRequest *req = r;
+        req->is_cancelled = true;
+
```

Comment:
This worries me a bit. There is no fundamental guarantee that the unref() caused by this is the final unref of the request object, so there is no guarantee that the  sd_bus_slot_unrefp() in the finalizer is run, to technically you may still reach agent_request_callback, and then the callback would be run twice.

I think we should either force-unref the bus slot, or check is_cancelled in agent_request_callback.
I think the later is maybe simpler.

I mean, right now this is only called on program quit, so its not gonna cause issues, but that may change later.

Reply:
Good point, done that